### ISSUE TITLE
CA2000 false positive when using fluent API's

### ISSUE THREAD

**Issue by lucastheisen** [dotnet/roslyn-analyzers#5709]
### Analyzer

**Diagnostic ID**: [CA2000](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2000): `Dispose objects before losing scope`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.401](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

I created a `Stream` wrapper for adding `IProgress` support.  It has fluent style registration for the reporters:

```csharp
        /// <summary>Registers a progress callback for calls to <c>Read</c>.</summary>
        /// <param name="progress">The callback, provided with the total bytes read to this
        /// point.</param>
        /// <returns>This stream for chaining.</returns>
        public ProgressStream AddReadProgressReporter(IProgress<long> progress)
        {
            this.readProgressCallbacks.Add(progress);
            return this;
        }

        /// <summary>Registers a progress callback for calls to <c>Write</c>.</summary>
        /// <param name="progress">The callback, provided with the total bytes written to this
        /// point.</param>
        /// <returns>This stream for chaining.</returns>
        public ProgressStream AddWriteProgressReporter(IProgress<long> progress)
        {
            this.writeProgressCallbacks.Add(progress);
            return this;
        }
```

However, when consumed using the fluent approach:

```csharp
using var stream = new ProgressStream(File.OpenRead("file.txt"))
    .AddWriteProgressReporter(reporter);
```

It will trigger this finding because it is not detecting that the stream returned by `.Add[Read|Write]ProgressReporter` is the same object.

I can work around it by separating into two commands:

```csharp
using var stream = new ProgressStream(File.OpenRead("file.txt"));
stream.AddWriteProgressReporter(reporter);
```

But this has other downsides.  I could ignore the warning at every consumption site, but that also has a huge downside to usability for consumers.

### Steps To Reproduce

1. Create a class that wraps a stream
2. Create a fluent API to modify the class and return `this`
3. Attempt to use in a fluent fashion

### Expected behavior

No warning

### Actual behavior

CA2000

**Comment by ltalasil24** [comment]
Is there any update on this issue? Would appreciate if any update can be provided regarding this. Is this issue fixed but the NuGet package isn't pushed yet as this seem to be open since 2021?

**Comment by bwilliams1** [comment]
I am also interested in an update on this issue. we have many false positives in our app that we would like cleaned up with the fix to this issue.

**Comment by ltalasil24** [comment]
Any update on this issue is appreciated. Thank you!

**Comment by sharwell** [comment]
I don't believe any fix is currently planned here. The current CA2000 behavior was largely derived from legacy FxCop. We've had a hard time making any alterations to the behavior without causing regressions (performance or functional) for some of our larger enterprise customers. It should be possible to implement a targeted `DiagnosticSuppressor` which understands how to automatically suppress CA2000 when it gets reported on one of your known fluent APIs.
